orte/tools/orterun/orterun.c

```diff
@@ -645,27 +646,67 @@ int orterun(int argc, char *argv[])
         return rc;
     }
 
+    /* print version if requested.  Do this before check for help so
+       that --version --help works as one might expect. */
+    if (orterun_globals.version) {
+        char *str, *project_name = NULL;
+        if (0 == strcmp(orte_basename, "mpirun")) {
+            project_name = "Open MPI";
+        } else {
+            project_name = "OpenRTE";
+        }
+        str = opal_info_make_version_str("all", 
+                                         OPAL_MAJOR_VERSION, OPAL_MINOR_VERSION, 
+                                         OPAL_RELEASE_VERSION, 
+                                         OPAL_GREEK_VERSION,
+                                         OPAL_REPO_REV);
+        if (NULL != str) {
+            fprintf(stdout, "%s (%s) %s\n\nReport bugs to %s\n",
+                    orte_basename, project_name, str, PACKAGE_BUGREPORT);
+            free(str);
+        }
+        exit(0);
+    }
+
+    /* DO NOT LET ROOT CALL ORTE_INIT/FINALIZE AS IT CAN BLAST SYSTEM FILES
+     * TO BE FULLY SAFE, WE DON'T ALLOW ANYTHING MORE THAN THE VERSION OUTPUT */
+    
     /* check if we are running as root - if we are, then only allow
      * us to proceed if the allow-run-as-root flag was given. Otherwise,
      * exit with a giant warning flag
      */
     if (0 == geteuid() && !orterun_globals.run_as_root) {
-        /* show_help is not yet available, so print an error manually */
-        fprintf(stderr, "--------------------------------------------------------------------------\n");
-        fprintf(stderr, "%s has detected an attempt to run as root. This is *strongly*\n", orte_basename);
-        fprintf(stderr, "discouraged as any mistake (e.g., in defining TMPDIR) or bug can\n");
-        fprintf(stderr, "result in catastrophic damage to the OS file system, leaving\n");
-        fprintf(stderr, "your system in an unusable state.\n\n");
-        fprintf(stderr, "You can override this protection by adding the --allow-run-as-root\n");
-        fprintf(stderr, "option to your cmd line. However, we reiterate our strong advice\n");
-        fprintf(stderr, "against doing so - please do so at your own risk.\n");
-        fprintf(stderr, "--------------------------------------------------------------------------\n");
+        if (orterun_globals.help) {
+            char *project_name = NULL;
+            if (0 == strcmp(orte_basename, "mpirun")) {
+                project_name = "Open MPI";
+            } else {
+                project_name = "OpenRTE";
+            }
+            fprintf(stderr, "%s cannot provide the help message when run as root\n"
+                    "Please run as regular user, or add the --run-as-root flag\n"
```

Comment:
Picky, picky - okay, I'll take care of both comments.


On Tue, Mar 31, 2015 at 1:28 PM, Jeff Squyres <notifications@github.com>
wrote:

> In orte/tools/orterun/orterun.c
> <https://github.com/open-mpi/ompi-release/pull/220#discussion_r27519689>:
>
> > -        fprintf(stderr, "discouraged as any mistake (e.g., in defining TMPDIR) or bug can\n");
> > -        fprintf(stderr, "result in catastrophic damage to the OS file system, leaving\n");
> > -        fprintf(stderr, "your system in an unusable state.\n\n");
> > -        fprintf(stderr, "You can override this protection by adding the --allow-run-as-root\n");
> > -        fprintf(stderr, "option to your cmd line. However, we reiterate our strong advice\n");
> > -        fprintf(stderr, "against doing so - please do so at your own risk.\n");
> > -        fprintf(stderr, "--------------------------------------------------------------------------\n");
> > +        if (orterun_globals.help) {
> > +            char *project_name = NULL;
> > +            if (0 == strcmp(orte_basename, "mpirun")) {
> > +                project_name = "Open MPI";
> > +            } else {
> > +                project_name = "OpenRTE";
> > +            }
> > +            fprintf(stderr, "%s cannot provide the help message when run as root\n"
> > +                    "Please run as regular user, or add the --run-as-root flag\n"
>
> I just noticed: the flag is actually --allow-run-as-root
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/open-mpi/ompi-release/pull/220/files#r27519689>.
>